metrics/src/tests.rs

```diff
@@ -0,0 +1,458 @@
+        }
+
+        Counter {
+            c: Ok(metric.get_counter()),
+            family_name: self.family.get_name().to_owned(),
+            metric_dump: self.registry.metrics_as_str(),
+        }
+    }
+
+    /// Returns the histogram metric, allowing assertions to be applied.
+    pub fn histogram(&mut self) -> Histogram<'_> {
+        // sort the assertion's labels
+        self.labels.sort_by(|a, b| a.0.cmp(&b.0));
+
+        let metric = self.family.get_metric().iter().find(|metric| {
+            if metric.get_label().len() != self.labels.len() {
+                return false; // this metric can't match
+            }
+
+            // sort this metrics labels and compare to assertion labels
+            let mut metric_labels = metric.get_label().to_vec();
+            metric_labels.sort_by(|a, b| a.get_name().cmp(b.get_name()));
+
+            // metric only matches if all labels are identical.
+            metric_labels
+                .iter()
+                .zip(self.labels.iter())
+                .all(|(a, b)| a.get_name() == b.0 && a.get_value() == b.1)
+        });
+
+        // Can't find metric matching labels
+        if metric.is_none() {
+            return Histogram {
+                c: NoMatchingLabelsError {
+                    name: self.family.get_name().to_owned(),
+                    labels: self.labels.clone(),
+                    metrics: self.registry.metrics_as_str(),
+                }
+                .fail(),
+                family_name: "".to_string(),
+                metric_dump: "".to_string(),
+            };
+        }
+        let metric = metric.unwrap();
+
+        if !metric.has_histogram() {
+            return Histogram {
+                c: FailedMetricAssertionError {
+                    name: self.family.get_name().to_owned(),
+                    msg: "metric not a counter".to_owned(),
+                    metrics: self.registry.metrics_as_str(),
+                }
+                .fail(),
+                family_name: "".to_string(),
+                metric_dump: "".to_string(),
+            };
+        }
+
+        Histogram {
+            c: Ok(metric.get_histogram()),
+            family_name: self.family.get_name().to_owned(),
+            metric_dump: self.registry.metrics_as_str(),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct Counter<'a> {
+    // if there was a problem getting the counter based on labels then the
+    // Error will contain details.
+    c: Result<&'a PromCounter, Error>,
+
+    family_name: String,
+    metric_dump: String,
+}
+
+impl<'a> Counter<'a> {
+    pub fn eq(self, v: f64) -> Result<(), Error> {
+        let c = self.c?; // return previous errors
+
+        if v != c.get_value() {
+            return FailedMetricAssertionError {
+                name: self.family_name,
+                msg: format!("{:?} == {:?} failed", c.get_value(), v),
+                metrics: self.metric_dump,
+            }
+            .fail();
+        }
```

Comment:
This (and similar future cases in this file) could probably use the snafu `ensure!` macro. Something like:

```suggestion
        ensure!(v == c.get_value(), FailedMetricAssertionError {
            name: self.family_name,
            msg: format!("{:?} == {:?} failed", c.get_value(), v),
            metrics: self.metric_dump,
        });            
```